contracts/LibSubmarineSimple.sol

```diff
@@ -0,0 +1,184 @@
+        uint128 amountRevealed,
+        uint128 amountUnlocked
+    ) {
+        CommitData memory sesh = commitData[_commitId];
+        return (
+            sesh.amountRevealed,
+            sesh.amountUnlocked
+        );
+    }
+
+    /////////////
+    // Setters //
+    /////////////
+
+    /**
+     * @notice Function called by the user to reveal the session.
+     * @dev warning Must be called within 256 blocks of the commit transaction to obtain the correct blockhash.
+     * @param _commitBlockNumber Block number in which the commit tx was included.
+     * @param _unlockAmount Value (i.e. how much money was sent) included in the commit tx.
+     * @param _embeddedDAppData optional Data passed embedded within the unlock tx. This should probably be null
+     * @param _witness Witness "secret" we committed to
+     * @param _unlockGasPrice Gas price of the unlock TX
+     * @param _unlockGasLimit Gas limit of the unlock TX
+     * @param _unlockTXHash full hash of the signed unlock transaction
+     * @param _proofBlob the Proof blob that gets passed to ethprove to verify merkle tree inclusion in a prior block.
+     */
+    function reveal(
+        uint32 _commitBlockNumber,
+        bytes _embeddedDAppData,
+        uint128 _unlockAmount,
+        bytes32 _witness,
+        uint256 _unlockGasPrice,
+        uint256 _unlockGasLimit,
+        bytes32 _unlockTXHash,
+        bytes _proofBlob
+    ) public {
+        bytes32 commitBlockHash = blockhash(_commitBlockNumber);
+        // fullCommit = (addressA + addressC + aux(sendAmount) + dappData + w + aux(gasPrice) + aux(gasLimit))
+
+        bytes32 commitId = getCommitId(
+            msg.sender,
+            address(this),
+            _unlockAmount,
+            _embeddedDAppData,
+            _witness,
+            _unlockGasPrice,
+            _unlockGasLimit
+        );
+
+        require(commitData[commitId].amountRevealed == 0, "The tx is already revealed");
+        require(commitBlockHash != 0x0, "Commit Block is too old to retreive block hash (more than 256 blocks), or does not exist");
+        require(block.number.sub(_commitBlockNumber) > commitPeriodLength, "You must wait long enough to allow the committing period to end before revealing");
+        ProvenTransaction memory proven_tx;
+        // Commented out for theoretical gas savings
+        (proven_tx.result, /* index */, proven_tx.nonce, /* gasprice */, /* startgas */, proven_tx.to, proven_tx.value, /* data */, /* v */ , /* r */, /* s */, proven_tx.is_contract_creation ) = txProof(commitBlockHash, _proofBlob);
+
+        // TX_PROOF_RESULT_PRESENT = 1;
+        // TX_PROOF_RESULT_ABSENT = 2;
+        require(proven_tx.result == 1, "The proof is invalid");
+        require(proven_tx.value >= _unlockAmount);
+        require(proven_tx.nonce == 0);
+        require(proven_tx.is_contract_creation == false);
+
+        address submarine = ecrecover(
+            _unlockTXHash,
```

Comment:
We just blindly accept `_unlockTxHash`, but we don't verify that `_unlockTx` has the right structure. We also don't check (in unlock) that the unlockTx is comming from the submarine address.

This means that the implementation is insecure because it enables an adversary to:
1. Generate an invalid "unlockTx" that returns funds to himself.
2. Send funds to the resulting invalid submarine address.
3. Refund himself.
4. Call reveal and unlock if he so pleases.

Am I overlooking something?

Reply:
The attack you describe should not be possible due to multiple mitigating controls:

line 127 uses the libsubmarine contract address as the receiver to generate the "Commit ID" - which is later then used to generate the submarine address to check against on lines 152 and 153. Attempting to call reveal will fail at during the require statement that checks the submarine address matches the recipient To: of the commit transaction. If someone wants to create an invalid unlock TX that returns funds to an attacker address instead of the LibSubmarine contract, they have to be able to generate a transaction that has R and S values that will still pass the reveal, which is breaking ECDSA.

https://github.com/lorenzb/libsubmarine/blob/f7b58ee4361105e5f85a967bb7b33569282ef2d9/contracts/LibSubmarineSimple.sol#L125-L127

https://github.com/lorenzb/libsubmarine/blob/f7b58ee4361105e5f85a967bb7b33569282ef2d9/contracts/LibSubmarineSimple.sol#L152-L153

Also, calling the unlock stores the quantity of funds sent in msg.value, so calling unlock requires fronting money regardless, though again, the mitigating control is intended to be in the reveal. Transaction spam to the unlock TX is 100% possible (by transaction spam I mean performing unlocks for random commitIDs) - though again, sending spam transactions requires gas money and even if there was a collision in keccak256, the full submarine workflow would still work.

 https://github.com/lorenzb/libsubmarine/blob/f7b58ee4361105e5f85a967bb7b33569282ef2d9/contracts/LibSubmarineSimple.sol#L169

Reply:
Nvm after discussion offline there is an attack here - an attacker creates an unlock and lies about the contents of the commit - i.e. creates a transaction where the commit(addressA, addressC, ... , witness, gas...) says it's pointing to the libsubmarine contract, but then the To: field of the transaction sends to another address entirely.

We're going to solve this by passing the RLP encoded unlock transaction as a parameter to the reveal, which allows us to also drop several variables in the reveal anyways - hopefully the gas costs don't go up too much.

Reply:
fixed in later commit